src/GovernorCountingFractional.sol

```diff
@@ -155,31 +183,45 @@ abstract contract GovernorCountingFractional is Governor {
      */
     function _countVoteFractional(
         uint256 proposalId,
+        ProposalVote storage _storedVote,
         address account,
         uint128 weight,
         bytes memory voteData
     ) internal {
         require(voteData.length == 48, "GovernorCountingFractional: invalid voteData");
 
-        (uint128 _forVotes, uint128 _againstVotes, uint128 _abstainVotes) = _decodePackedVotes(voteData);
-
-        uint128 _existingWeight = _proposalVotersWeightCast[proposalId][account];
-        uint256 _newWeight = uint256(_forVotes) + _againstVotes + _abstainVotes + _existingWeight;
+        ProposalVote memory _newVote = _decodePackedVotes(voteData);
+        uint256 _newWeight = _totalProposalVoteWeights(_newVote);
 
         require(_newWeight <= weight, "GovernorCountingFractional: vote would exceed weight");
 
-        // It's safe to downcast here because we've just confirmed that
-        // _newWeight < weight, and we know that weight is <= uint128.max.
-        _proposalVotersWeightCast[proposalId][account] = uint128(_newWeight);
-
-        ProposalVote memory _proposalVote = _proposalVotes[proposalId];
-        _proposalVote = ProposalVote(
-            _proposalVote.againstVotes + _againstVotes,
-            _proposalVote.forVotes + _forVotes,
-            _proposalVote.abstainVotes + _abstainVotes
+        require(_storedVote.againstVotes <= _newVote.againstVotes, "cannot undo past votes");
+        require(_storedVote.forVotes <= _newVote.forVotes, "cannot undo past votes");
+        require(_storedVote.abstainVotes <= _newVote.abstainVotes, "cannot undo past votes");
+
+        // How many votes are we adding here?
+        ProposalVote memory _voteDelta;
```

Comment:
Why do we calculate the delta instead of just overwriting?

Reply:
Good question!

Because the previous votes have already been added to the overall voting tallies. To make sure that we don't double count votes, we only increment the overall tallies by the delta between what was cast before and what is being cast now.